README.md
# SSO-自有系統前端 SSO 封裝(SDK) Ver 2.0

# 主要

```text
dist/
├── cross-token-access.cjs.js (CommonJS)
├── cross-token-access.cjs.min.js (CommonJS, compressed)
├── cross-token-access.umd.js (UMD)
└── cross-token-access.umd.min.js (UMD, compressed)
```

# 依賴 & 注意事項

依賴：

- **axios**
- **lodash**
- **js-cookie**
- **promise.prototype.finally**

注意事項：

- 依賴套件已打包封裝，無需再外部引用
- API 採用 Axios 技術
- Axios 支援 finally，引用 promise.prototype.finally 方法
- Axios 攔截器 (interceptor) 有加入請求重複發送機制

# 跨域信任域名

- `nueip.com`
- `nueip.site`

# 使用說明

## 安裝

Script tag

```html
<script src="your-path/cross-token-access.umd.min.js"></script>
```

npm

```bash
npm install @nueip/cross-token-access
```

## 用法

```js
// 實體化
var tokenInjection = new TokenInjection(options);
```

- **options**
  - Type: `Object`
  - 參考 [options](#options)

## 範例

請見[範例](./examples/index.html)

# Options

### autopilot

- Type: `Boolean`
- Default: `true`
- note: 是否自動初始化 & 刷新同步

### sso_url

- Type: `String`
- Default: `empty string`
- note: SSO 伺服器 URL

### cookie_prefix

- Type: `String`
- Default: `empty string`
- note: Cookie 自定義前綴字串

### redirect_url

- Type: `String`
- Default: `empty string`
- note: 重新定向網址

### xhr_with

- Type: `Boolean`
- Default: `false`
- note: 是否配置 X-Requested-With 抬頭

### onLogout

- Type: `Function`
- Default: `null`
- note: 非登入狀態的回調函式

### unauthorized

- Type: `Function`
- Default: `null`
- note: 401 未授權的 Callback

### maxSyncTimesPerMinute

- Type: `number`
- Default: 3
- note: 每分鐘同步 token 最大次數限制

### maxRefreshTimesPerMinute

- Type: `number`
- Default: 3
- note: 每分鐘刷新 token 最大次數限制

# Methods

## init()

- 初始化執行
  - 自動同步/刷新 Token
  - 自動登出倒數

## sync()

- return

  - Promise 方法

- 執行一次，向 oAuth Server 同步 Token 資訊，同步錯誤時，檢查是否為登入狀態，否時刪除 Token

## autoSync(interval)

- interval
  - Type: `Number`
  - Note: 多少個間隔，每個間為 1000 毫秒
- 定期執行，向 oAuth Server 同步 Token 資訊
- 執行條件
  - Cookie 中 tkchecksum 是否與 LocalStorage 中的 token_checksum 不一樣
  - Axios 未執行過或已執行完成
- 多視窗時有可能同時執行，待觀察
- 請求次數超過，自動登出
- 執行錯誤時關閉自動同步 3 秒後重啟

## autoSyncStop()

- 停止自動同步 Token 內容

## refresh()

- return
  - Promise 方法
- throw
  - 沒有 Refresh Token 時丟出
- 執行一次，向 oAuth Server 執行 Refresh Token
- 執行條件
  - 必需有 refresh_token 金鑰：localStorage.token_refresh_token
  - 當現在時間超過到期時間 - TokenRefreshBefore 時觸發更新 token

## autoRefresh(interval)

- interval
  - Type: `Number`
  - Note: 多少個間隔，每個間為 1000 毫秒
- 定期執行，向 oAuth Server 同步 Token 資訊
- 執行條件
  - 即將過期
  - Axios 未執行過或已執行完成
- 多視窗時有可能同時執行，待觀察
- 請求次數超過，自動登出
- 執行錯誤時關閉自動同步 3 秒後重啟

## autoRefreshStop

- 停止自動刷新 Token

## getToken()

- return
  - Type: `String`
  - Local Storage 中的 Token
- 取得 Local Storage Token

## getLang()

- return
  - Type: `String`
  - Cookie 中的語系代號
- 取得當前語系

## validate(token)

驗證 Token

- token
  - Type: `String`
  - 本地端要被驗證的 Token
- return
  - Promise 方法

## loginIAM()

- 開啟 IAM 登入頁面，另開新視窗
- 轉導自動帶入GET參數 redirect_uri={指定重新定向網址}

## logoutIAM()

- 開啟 IAM 登出頁面

## isLogin()

- return
  - Boolean
- 判斷是否為登入狀態

## axiosCreate()

- config
  - Type: `Object`
  - axios 請求參數
  - 只支援 baseURL, withCredentials, timeout, headers
- return
  - Boolean
- axios 全域設定輸入接口

# 瀏覽器支援

- ### Chrome (latest)

- ### Firefox (latest)

- ### Safari (latest)

- ### Edge (latest)

- ### ~~Internet Explorer 11~~


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.3.0](https://github.com/nueip/cross-token-access/compare/v2.2.3...v2.3.0) (2023-12-13)


### Features

* 擴充取得子系統可用清單方法 ([40da2df](https://github.com/nueip/cross-token-access/commit/40da2df81b4b4f6b501033b97cd84e5f5ab0057f))


### Bug Fixes

* **sonarlint:** 修正邏輯語法 ([e81b406](https://github.com/nueip/cross-token-access/commit/e81b4063ad0790c5b2723ddda17bbe276372090d))

### [2.2.3](https://github.com/nueip/cross-token-access/compare/v2.2.2...v2.2.3) (2023-05-26)


### Bug Fixes

* 調整登入狀態判斷 ([a853a8a](https://github.com/nueip/cross-token-access/commit/a853a8a6c1b005a6eab6a3319d995aa66d649203))

### [2.2.2](https://github.com/nueip/cross-token-access/compare/v2.2.1...v2.2.2) (2023-05-08)


### Bug Fixes

* 修正 `reset()` 清除定期器邏輯 ([09a877e](https://github.com/nueip/cross-token-access/commit/09a877e79bd0e2513d4d49b716926662162bb50e))

### [2.2.1](https://github.com/nueip/cross-token-access/compare/v2.2.0...v2.2.1) (2023-05-03)

## [2.2.0](https://github.com/nueip/cross-token-access/compare/v2.1.10...v2.2.0) (2023-05-03)


### Features

* **lib:** 增加 `isSet()` 函式 ([15edea0](https://github.com/nueip/cross-token-access/commit/15edea01abe59ad6c5d84052839fd1f6fbddfced))
* 增加每分鐘同步/刷新 token 次數限制 ([8af0df0](https://github.com/nueip/cross-token-access/commit/8af0df04185929aa92aebde9d1ac0d5757e7f8af))


### Bug Fixes

* 修正 `autoRefresh()` 函式 ([d62b272](https://github.com/nueip/cross-token-access/commit/d62b2721428273c3f671a384be2f272e44b837f8))
* 修正 autoSync 同步回應狀態判斷 ([2348abc](https://github.com/nueip/cross-token-access/commit/2348abc09f072f3c522253f85977002fbe815d48))
* 修正 autoSync 檢核碼檢查函式 ([6960c18](https://github.com/nueip/cross-token-access/commit/6960c184f4993e195b9058137dc2cc7ea82744ca))
* 定期同步/刷新 token 重啟前先佔用，避免重啟期間再次執行函式時誤判問題 ([66f9957](https://github.com/nueip/cross-token-access/commit/66f9957c17dff814b41dcfe3179df2c5a9b1489c))

### 2.1.10(2022-12-15)
### Patch

* Improve `cookie_prefix` string  ([d280483](https://github.com/nueip/cross-token-access/pull/47/commits/d280483c058323c44f34937775570687679fe91b))

### 2.1.9 (2022-12-14)
### Patch

* Update package ([eafc5e5](https://github.com/nueip/cross-token-access/pull/46/commits/eafc5e5c1626e062a30e12d532f338e3d04dbcaf))

* Sync Initialization - Redirect to IAM logout page when logout status changes  ([4da5f16](https://github.com/nueip/cross-token-access/pull/46/commits/4da5f169636e31fbdc915e1979b9a3933396e5b4))

### 2.1.8 (2022-10-13)
### Patch

* Update package ([000f19b](https://github.com/nueip/cross-token-access/commit/b76750ebf4e3aca93917d393818898b0eeeb36ca))

### 2.1.7 (2022-10-13)
### Patch

* Update package ([000f19b](https://github.com/nueip/cross-token-access/commit/478e88009678f3c06ff9a4f3520eb4a7505cd010))

* Uninstall `lodash` and Optimize APIs url  ([f9cb2db](https://github.com/nueip/cross-token-access/commit/980180cdfa594f0c24d81b8aa17ae1f235b60730))

* Auto sync exclude 401 error  ([f5456d1](https://github.com/nueip/cross-token-access/commit/8e82247813660d972c06e1dc5dab35577f6e1d8c))

### 2.1.4 (2022-04-15)
### Patch

* Update package and improve rollup config  ([000f19b](https://github.com/nueip/cross-token-access/commit/000f19b135576bd664e4467c3aeb106983f7a9c4))

* Improve `cookie_prefix` options  ([f9cb2db](https://github.com/nueip/cross-token-access/commit/f9cb2db21f1e4966a2d214ff7bd3d2dca8433af7))

* Use `Promise` object to refresh token and validate token APIs  ([f5456d1](https://github.com/nueip/cross-token-access/commit/f5456d17f13f643debc7195bc8bc4214de13bb83))

### 2.1.2 (2022-04-08)
### Patch

* Changed `LOGOUT_TIME` to 12 hours ([c64a8fb8](https://github.com/nueip/cross-token-access/commit/c64a8fb8bbeefc6b9b6d67dc11dea223fd85a713))

* Optimize Eslint format and variable declaration ([b6ddc72c](https://github.com/nueip/cross-token-access/commit/b6ddc72c65a5400d63ca6da081cf87fcf5fff768))


### 2.1.1 (2022-03-03)
### Doc

* Add `getLang()` method usage

### 2.1.0 (2022-03-01)
### Features

* Add `getLang()` method ([468f7abd](https://github.com/nueip/cross-token-access/commit/468f7abdddb7ab818c5e23165ea22846bdc1299b))


src/token-injection.js
/**
 * Token Injection Client side
 *
 * @author Grace.Wang
 */
import cookies from 'js-cookie';
import * as TC from './constant';
import { rand, deepMerge, queryString, isSet } from './lib';
import { api, httpRequest } from './helpers/request';
import { setTokens, removeTokens } from './helpers/token';
import errorMsg from './helpers/error-message';
import webStorage from './helpers/storage';
import privateMethods from './privateMethods';
import RateLimitError from './errors/rate-limit-error';

// Axios 支援 finally 方法
require('promise.prototype.finally').shim();

// 初始預設值
const DEFAULTS = {
  // 是否自動初始化 & 刷新同步
  autopilot: true,
  // 單一登入網址
  sso_url: '',
  // 自定義 Cookie 前綴字串
  cookie_prefix: '',
  // 重新定向網址
  redirect_url: '',
  // 是否配置 X-Requested-With 抬頭
  xhr_with: true,
  // 非登入狀態的 Callback
  onLogout: null,
  // 401 未授權的 Callback
  unauthorized: null,
  // 每分鐘同步 token 最大次數限制
  maxSyncTimesPerMinute: 3,
  // 每分鐘刷新 token 最大次數限制
  maxRefreshTimesPerMinute: 3,
};

class TokenInjection {
  /**
   * 建構子
   *
   * @param {object} options - 參數設定
   */
  constructor(options = {}) {
    // 選項屬性
    this.options = deepMerge(DEFAULTS, options);

    // Token Keys
    this.tokenKeys = [
      TC.ACCESS_TOKEN_NAME,
      TC.TOKEN_EXPIRED_NAME,
      TC.TOKEN_TYPE,
      TC.TOKEN_SCOPE,
      TC.REFRESH_TOKEN_NAME,
      TC.TOKEN_CHECK_SUM,
      TC.TOKEN_CREATE_TIME_NAME,
    ];

    // Schedule cache
    this.intervalSync = null;
    this.intervalRefresh = null;

    // 自動同步/刷新 Token 請求錯誤次數計數
    this.syncTimes = 0;
    this.refreshTimes = 0;

    // 初始化同步 token 計數、刷新 token 計數
    this.syncCount = 0;
    this.refreshCount = 0;

    // 定期每 60 秒 (1 分鐘) 重置計數
    setInterval(() => {
      this.syncCount = 0;
      this.refreshCount = 0;
    }, 60 * 1000);

    // 暫存執行中的請求資訊
    this.axiosPending = new Map();

    // 實例化 axios
    this.rest = httpRequest({
      baseURL: this.options.sso_url,
      // 判斷是否為 Ajax 非同步請求，跨域時須自行配置此 header access
      headers: { 'X-Requested-With': 'XMLHttpRequest' },
    });

    // Axios 攔截器
    privateMethods.interceptors(this);

    // 自動初始化
    if (this.options.autopilot) {
      this.init();
    }
  }

  /**
   * 初始化 TokenInjection 實例
   *
   * @returns {Promise}
   */
  async init() {
    const instance = this;

    return instance
      .sync()
      .then(() => {
        // 載入後執行 定期同步 Token 內容
        instance.autoSync();

        // 載入後執行 定期刷新 Token
        instance.autoRefresh();

        // 載入後執行 自動登出倒數
        privateMethods.autoLogout(instance);
      })
      .catch((error) => {
        // 捕獲錯誤為登出狀態，轉導回 IAM 登出頁
        if (error.isLogout) instance.logoutIAM();

        // rate limit error 處理
        if (error instanceof RateLimitError) {
          throw error;
        }

        throw new Error(error);
      });
  }

  /**
   * 同步 Token 內容 - 執行一次
   *
   * - 向 oAuth Server 同步 Token 資訊
   * - 同步錯誤時，檢查是否為登入狀態，否時刪除 Token
   *
   * @returns {object}
   */
  sync() {
    const instance = this;
    const { rest, tokenKeys, options } = this;

    // 同步 Token 計數 + 1
    instance.syncCount += 1;

    // 檢查是否超過每分鐘同步 token 最大次數限制
    if (instance.syncCount > options.maxSyncTimesPerMinute) {
      return Promise.reject(new RateLimitError('sync'));
    }

    return new Promise((resolve, reject) => {
      // 抓取資料
      rest
        .get(`${options.sso_url}${api.sync}`)
        .then((res) => {
          let tokenInfo = res.data || {}; //eslint-disable-line

          // 執行完成，暫存請求響應狀態
          instance.axiosPending.set('sync', res.request.readyState);

          // 請求次數計數歸零
          instance.syncTimes = 0;

          // 預設清除 Token Keys
          removeTokens(tokenKeys);

          // 寫入 Token Keys 異常，清空 localStorage 後，執行登出
          try {
            setTokens(tokenKeys, tokenInfo);
          } catch (error) {
            webStorage.clear();
            instance.logoutIAM();
          }

          resolve(res);
        })
        .catch((error) => {
          // 請求次數計數
          instance.syncTimes += 1;

          // 請求次數超過最大限制錯誤訊息
          if (instance.syncTimes >= TC.MAX_REQUEST_TIMES) {
            console.error(`[Sync]: ${errorMsg.maxRequest}`);
          }

          reject(error);
        });
    });
  }

  /**
   * 刷新 Token - 執行一次
   *
   * - 向 oAuth Server 執行 Refresh Token
   * - 執行條件
   * - 必需有 refresh_token 金鑰: localStorage.token_refresh_token
   * - 當 現在時間 超過 過期時間 - TokenRefreshBefore 時觸發更新 token
   *
   * @throws 沒有 Refresh Token 時丟出例外
   * @returns {Promise}
   */
  refresh() {
    const instance = this;
    const { rest, options } = this;

    // Refresh Token 值
    let refreshToken = webStorage.get(TC.REFRESH_TOKEN_NAME); //eslint-disable-line

    // 金鑰不存在時丟出例外
    if (!refreshToken) {
      throw privateMethods.exception(instance, 'Need Refresh Token !', 401);
    }

    // 刷新 Token 計數 + 1
    instance.refreshCount += 1;

    // 檢查是否超過每分鐘刷新 token 最大次數限制
    if (instance.refreshCount > options.maxRefreshTimesPerMinute) {
      return Promise.reject(new RateLimitError('refresh'));
    }

    // 執行刷新金鑰
    return new Promise((resolve, reject) => {
      rest
        .post(
          `${options.sso_url}${api.refresh}?v=${rand(11111, 99999)}`,
          queryString({ refresh_token: refreshToken }),
          {
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
          }
        )
        .then((res) => {
          // 請求次數計數歸零
          instance.refreshTimes = 0;

          // 執行完成，暫存請求響應狀態
          instance.axiosPending.set('refresh', res.request.readyState);

          resolve(res);
        })
        .catch((error) => {
          // 請求次數計數
          instance.refreshTimes += 1;

          // 請求次數超過最大限制錯誤訊息
          if (instance.refreshTimes >= TC.MAX_REQUEST_TIMES) {
            console.error(`[Refresh]: ${errorMsg.maxRequest}`);
          }

          reject(error);
        });
    });
  }

  /**
   * 定期執行同步 Token 內容
   *
   * - 向 oAuth Server 同步Token資訊
   * - 執行條件
   * - Cookie 中 tkchecksum 是否與 LocalStorage 中的 token_checksum 不一樣
   * - axios未執行過或已執行完成
   * - 多視窗時有可能同時執行，待觀察
   * - 執行錯誤時關閉自動同步30秒後重啟
   *
   * @param {number} interval - 多少個間隔，每個間隔為1分鐘
   */
  autoSync(interval = 0) {
    const instance = this;
    const { options } = this;
    const tkCheckSumCookieName = `${options.cookie_prefix}tkchecksum`;
    // 檢查 LocalStorage 金鑰檢核碼與 Cookie 金鑰檢核碼是否一致
    const validateChecksum = () => {
      const checksumFromCookie = cookies.get(tkCheckSumCookieName);
      const checksumFromLocalStorage = webStorage.get('token_checksum');

      // 若 checksum 未設定則回傳 false 表示不一致 (避免皆為 null 或 undefined 時誤判)
      if (!isSet(checksumFromCookie, checksumFromLocalStorage)) {
        return false;
      }

      return checksumFromCookie === checksumFromLocalStorage;
    };

    // 若已定期執行則中斷處理
    if (instance.intervalSync) {
      return;
    }

    // 定期執行 (Cookie 中的金鑰檢核碼必須存在)
    instance.intervalSync = setInterval(async () => {
      // 若 tkchecksum 與 token_checksum 一致 或 正在處理同步中，則略過本次處理
      if (validateChecksum() || instance.isProcessing('sync')) {
        return;
      }

      await instance.sync().catch((error) => {
        // 取得 回覆資源
        const { response } = error;
        // 取得 錯誤狀態碼
        let errorCode = response ? response.status : 0; //eslint-disable-line

        // 排除 401 Code：Token 失效發還狀態
        if (errorCode === 401) {
          return;
        }

        // rate limit error 處理
        if (error instanceof RateLimitError) {
          console.error(error);
        }

        // 執行錯誤時關閉自動同步 等待30秒鐘後重啟
        instance.autoSyncStop();
        // 重啟前先占用，避免 30 內重啟期間內再次執行 autoSync 時誤判，導致重複建立定期同步 Token
        instance.intervalSync = -1;
        setTimeout(() => {
          // 檢查非占用狀態則不處理 (排除已重新執行 autoSync 或執行後又再停止的情況)
          if (instance.intervalSync !== -1) {
            return;
          }

          // 清除占用
          instance.intervalSync = null;
          // 重啟定期執行同步 Token
          instance.autoSync();
        }, TC.TOKEN_AUTO_SYNC_RESTART);
      });
    }, 1000 * 60 * Math.abs(interval) || TC.TOKEN_AUTO_SYNC_INTERVAL);
  }

  /**
   * 停止 自動同步 Token 內容
   */
  autoSyncStop() {
    const instance = this;

    if (instance.intervalSync) {
      // 停止定期執行
      clearInterval(instance.intervalSync);
      instance.intervalSync = null;
      instance.axiosPending.delete('sync');
    }
  }

  /**
   * 定期刷新 Token
   *
   * - 向 oAuth Server 同步Token資訊
   * - 執行條件
   * - 即將過期
   * - axios未執行過或已執行完成
   * - 多視窗時有可能同時執行，待觀察
   * - 執行錯誤時關閉自動同步3秒後重啟
   *
   * @param {Number} interval - 多少秒
   */
  autoRefresh(interval = 0) {
    const instance = this;

    // 執行錯誤時關閉自動同步30秒後重啟
    const restartAutoRefresh = () => {
      instance.autoRefreshStop();
      // 重啟前先占用，避免 30 內重啟期間內再次執行 autoRefresh 時誤判，導致重複建立定期刷新 Token
      instance.intervalRefresh = -1;
      setTimeout(() => {
        // 檢查非占用狀態則不處理 (排除已重新執行 autoRefresh 或執行後又再停止的情況)
        if (instance.intervalSync !== -1) {
          return;
        }

        // 清除占用
        instance.intervalRefresh = null;
        // 重啟定期刷新 Token
        instance.autoRefresh();
      }, 30000);
    };

    // 已定期執行則中斷處理
    if (instance.intervalRefresh) {
      return;
    }

    instance.intervalRefresh = setInterval(() => {
      try {
        // 若 refresh request 正在處理，則略過本次處理
        if (instance.isProcessing('refresh')) {
          return;
        }

        // 現在時間
        const nowTime = parseInt(Date.now() / 1000, 10);

        // Token建立時間
        const createKey = webStorage.get(TC.TOKEN_CREATE_TIME_NAME);
        const createTime = parseInt(createKey, 10);

        // Token過期時間
        const expiredKey = webStorage.get(TC.TOKEN_EXPIRED_NAME);
        const expireTime = parseInt(expiredKey, 10);

        // 過期時間 - TokenRefreshBefore
        const refreshTime = createTime + expireTime - TC.TOKEN_REFRESH_BEFORE;

        // 當 現在時間 超過 過期時間 - TokenRefreshBefore 時觸發更新 Token
        if (nowTime >= refreshTime) {
          instance.refresh().catch((error) => {
            // rate limit error 處理
            if (error instanceof RateLimitError) {
              console.error(error);
            }
            // 執行錯誤時關閉自動同步30秒後重啟
            restartAutoRefresh();
          });
        }
      } catch (e) {
        // 例外訊息
        console.error(`[${e.code}]${e.message}`);
        restartAutoRefresh();
      }
    }, interval * 1000 || TC.TOKEN_AUTO_REFRESH_INTERVAL * 1000);
  }

  /**
   * 停止 自動刷新 Token
   */
  autoRefreshStop() {
    const instance = this;

    if (instance.intervalRefresh) {
      // 停止定期執行
      clearInterval(instance.intervalRefresh);
      instance.intervalRefresh = null;
      instance.axiosPending.delete('refresh');
    }
  }

  /**
   * 驗證Token
   *
   * @param {string} token - 本地端要被驗證的 Token
   * @returns {Promise}
   */
  validate(token) {
    const { rest, options } = this;
    let validateToken = token || ''; //eslint-disable-line

    return new Promise((resolve, reject) => {
      rest
        .get(`${options.sso_url}${api.validate}?v=${rand(11111, 99999)}`, {
          headers: {
            Authorization: `Bearer ${validateToken}`,
          },
        })
        .then((res) => {
          resolve(res);
        })
        .catch((error) => {
          reject(error);
        });
    });
  }

  /**
   * 取得 LocalStorage Token
   *
   * @returns {string} Access Token
   */
  getToken() {
    return webStorage.get(TC.ACCESS_TOKEN_NAME);
  }

  /**
   * 取得當前語系
   *
   * @returns {string} 回傳語系代碼
   */
  getLang() {
    const { options } = this;
    return cookies.get(`${options.cookie_prefix}lang`) || 'en';
  }

  /**
   * 取得可用子系統清單
   *
   * @returns {Array} 回傳系統清單
   */
  getSubSystem() {
    const subList = cookies.get('login')?.split('.') ?? [];
    const subSystemSort = ['hrm', 'bpm', 'crm', 'sales', 'bi'];

    return subSystemSort.filter((subSystem) => subList.includes(subSystem));
  }

  /**
   * 開啟登入頁面
   *
   * @param {string} target - _self | _blank
   */
  loginIAM(target = '') {
    const { options } = this;
    const ssoUrl = `${options.sso_url}/login?redirect_uri=${options.redirect_url}`;

    window.open(ssoUrl, target || '_self');
  }

  /**
   * 登出
   */
  logoutIAM() {
    const instance = this;
    const { options } = this;
    const ssoUrl = `${options.sso_url}/logout`;

    // 重置初始建構屬性 & 清除 Token's Info.
    privateMethods.reset(instance).then(() => {
      // 轉導回 SSO 登出頁
      window.location.href = ssoUrl;
    });
  }

  /**
   * 是否為登入狀態
   *
   * @returns {Boolean}
   */
  isLogin() {
    const { options } = this;
    const loginKey = `${options.cookie_prefix}login`;
    const loginCookie = cookies.get(loginKey);

    return Boolean(loginCookie);
  }

  /**
   * 檢查目標 requestName 的 readyState 狀態是否正在處理中
   *
   * @param {string} requestName 'sync', 'refresh'
   * @returns {boolean}
   */
  isProcessing(requestName) {
    const requestReadyState = this.axiosPending.get(requestName);

    return isSet(requestReadyState) && requestReadyState !== 4;
  }

  /**
   * axios 全域設定方法
   *
   * @param {object} config - axios options
   * @returns {TokenInjection}
   */
  axiosCreate(config = {}) {
    const instance = this;

    instance.options.sso_url = config.baseURL || '';
    instance.rest = httpRequest(config);

    return instance;
  }
}

export default TokenInjection;
